# Analysis for LEAVETIMES database  : Dublin Bus

* leavetimes database is vast. For representative purpose, we shall start with data for month of january

In [1]:
import pandas as pd

## Vehicles data

### Load file into dataframe

In [2]:
df_Jan = pd.read_csv("../DB/monthlyData/January_2018.csv",skip_blank_lines=True,index_col=False)

### View vehicle data

In [3]:
df_Jan.head()

DAYOFSERVICE   TRIPID  PROGRNUMBER  STOPPOINTID  PLANNEDTIME_ARR  \
0   2018-01-01  5958355            3         1172            23799   
1   2018-01-01  5958355            4         1173            23860   
2   2018-01-01  5958355            5         1174            23937   
3   2018-01-01  5958355            6         1175            24048   
4   2018-01-01  5958355            7           15            24130   

   PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP  VEHICLEID  
0            23799           23863           23863    1001237  
1            23860           23934           24072    1001237  
2            23937           24114           24140    1001237  
3            24048           24180           24192    1001237  
4            24130           24227           24257    1001237

### Properties of features

In [4]:
df_Jan.info(null_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10525004 entries, 0 to 10525003
Data columns (total 9 columns):
 #   Column           Non-Null Count     Dtype 
---  ------           --------------     ----- 
 0   DAYOFSERVICE     10525004 non-null  object
 1   TRIPID           10525004 non-null  int64 
 2   PROGRNUMBER      10525004 non-null  int64 
 3   STOPPOINTID      10525004 non-null  int64 
 4   PLANNEDTIME_ARR  10525004 non-null  int64 
 5   PLANNEDTIME_DEP  10525004 non-null  int64 
 6   ACTUALTIME_ARR   10525004 non-null  int64 
 7   ACTUALTIME_DEP   10525004 non-null  int64 
 8   VEHICLEID        10525004 non-null  int64 
dtypes: int64(8), object(1)
memory usage: 722.7+ MB


In [5]:
df_Jan.nunique()

DAYOFSERVICE          31
TRIPID             76453
PROGRNUMBER          103
STOPPOINTID         4714
PLANNEDTIME_ARR    68859
PLANNEDTIME_DEP    68859
ACTUALTIME_ARR     71982
ACTUALTIME_DEP     71965
VEHICLEID            996
dtype: int64

**REVIEW OF FEATURES**

| Features        | Observation   | Decision  |
| :--------------- |:-------------|:-----|
| **DAYOFSERVICE**      | Non -null column with 31 value; represent each day of month| **No action** |
| **TRIPID**    | Non -null column with  76453 unique entries; indicates 1 way trip made by bus on defined route| **No action** |
| **PROGRNUMBER**       | Non -null column with  103 unique entries; indicate position of stop in trip| **No action** |
| **STOPPOINTID**        | Non -null column with  4714 unique entries; Nominal catagorical variable indicating stop number| **No action** |
| **PLANNEDTIME_ARR**    | Non -null column with  68859 unique entries; expresses defined time in seconds [offset from 00:00:00 of the day] for bus to arrive at a stop| **No action** |
| **PLANNEDTIME_DEP**      | Non -null column with  68859 unique entries; expresses defined time in seconds [offset from 00:00:00 of the day] for bus to depart from a stop| **No action** |
| **ACTUALTIME_ARR**       | Non -null column with  71982 unique entries; expresses defined time in seconds [offset from 00:00:00 of the day] for bus to arrive at a stop| **No action** |
| **ACTUALTIME_DEP**       | Non -null column with  71965 unique entries; expresses defined time in seconds [offset from 00:00:00 of the day] for bus to depart from a stop| **No action** |
| **VEHICLEID**            | Non -null column with  996 unique entries;  expresses perticular vehicle used for trip, NOT SIGNIFICANT FOR ANALYSIS| **Drop** |

In [6]:
# drop VEHICLEID feature
df_Jan.drop('VEHICLEID',axis=1,inplace=True)

## View data for perticular trip ID

Considering that we are totally unaware of dublin bus operations; let us check assumptions:
* All buses are operational on all 31 days of service [hypothesis for fully loaded functional service]
* Each one way trip has unique ID [Established trueth by documentation]
* Thus, we may predict that :<br>
$Average\ daily\ trips\ for\ a\ bus=\dfrac{trips\ per\ day}{Number\ of \ buses}$
$\dfrac{76453/31}{996} = 2.47$


### Convert DAYOFSERVICE to datetime 

In [7]:
df_trip_Jan = df_Jan.copy() 
df_trip_Jan['DAYOFSERVICE'] =  pd.to_datetime(df_trip_Jan['DAYOFSERVICE'],format='%Y-%m-%d')
df_trip_Jan.head()

DAYOFSERVICE   TRIPID  PROGRNUMBER  STOPPOINTID  PLANNEDTIME_ARR  \
0   2018-01-01  5958355            3         1172            23799   
1   2018-01-01  5958355            4         1173            23860   
2   2018-01-01  5958355            5         1174            23937   
3   2018-01-01  5958355            6         1175            24048   
4   2018-01-01  5958355            7           15            24130   

   PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP  
0            23799           23863           23863  
1            23860           23934           24072  
2            23937           24114           24140  
3            24048           24180           24192  
4            24130           24227           24257

In [8]:
df_trip_Jan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10525004 entries, 0 to 10525003
Data columns (total 8 columns):
 #   Column           Dtype         
---  ------           -----         
 0   DAYOFSERVICE     datetime64[ns]
 1   TRIPID           int64         
 2   PROGRNUMBER      int64         
 3   STOPPOINTID      int64         
 4   PLANNEDTIME_ARR  int64         
 5   PLANNEDTIME_DEP  int64         
 6   ACTUALTIME_ARR   int64         
 7   ACTUALTIME_DEP   int64         
dtypes: datetime64[ns](1), int64(7)
memory usage: 642.4 MB


### Visualise data for perticular trip 

In [9]:
TRIPID_unique = df_trip_Jan['TRIPID'].unique()
group_TRIPID = df_trip_Jan.groupby(by='TRIPID',axis=1)

In [10]:
df_trip_Jan.loc[(df_trip_Jan['PLANNEDTIME_ARR'] == 30600) & (df_trip_Jan['STOPPOINTID'] == 324)]

DAYOFSERVICE   TRIPID  PROGRNUMBER  STOPPOINTID  PLANNEDTIME_ARR  \
1414       2018-01-01  5959280            1          324            30600   
247416     2018-01-02  5970229            1          324            30600   
247455     2018-01-02  5967072            1          324            30600   
618693     2018-01-03  5970229            1          324            30600   
618766     2018-01-03  5967072            1          324            30600   
998788     2018-01-04  5970229            1          324            30600   
998830     2018-01-04  5967072            1          324            30600   
1380013    2018-01-05  5970229            1          324            30600   
1380070    2018-01-05  5967072            1          324            30600   
1738227    2018-01-06  5970892            1          324            30600   
2004978    2018-01-07  6099430            1          324            30600   
2252962    2018-01-08  6096749            1          324            30600   
2252981    2018-01-08  6091671            1          324            30600   
2635900    2018-01-09  6096749            1          324            30600   
2635904    2018-01-09  6091671            1          324            30600   
3014535    2018-01-10  6096749            1          324            30600   
3014554    2018-01-10  6091671            1          324            30600   
3393582    2018-01-11  6091671            1          324            30600   
3393633    2018-01-11  6096749            1          324            30600   
3769414    2018-01-12  6096749            1          324            30600   
3769442    2018-01-12  6091671            1          324            30600   
4128791    2018-01-13  6098996            1          324            30600   
4402249    2018-01-14  6120466            1          324            30600   
4653566    2018-01-15  6114321            1          324            30600   
4653575    2018-01-15  6116449            1          324            30600   
5033477    2018-01-16  6116449            1          324            30600   
5033506    2018-01-16  6114321            1          324            30600   
5413422    2018-01-17  6114321            1          324            30600   
5791721    2018-01-18  6116449            1          324            30600   
5791762    2018-01-18  6114321            1          324            30600   
6172089    2018-01-19  6116449            1          324            30600   
6172117    2018-01-19  6114321            1          324            30600   
6530647    2018-01-20  6121309            1          324            30600   
6800627    2018-01-21  6227609            1          324            30600   
7051674    2018-01-22  6227874            1          324            30600   
7051682    2018-01-22  6214488            1          324            30600   
7431219    2018-01-23  6214488            1          324            30600   
7431226    2018-01-23  6227874            1          324            30600   
7807915    2018-01-24  6214488            1          324            30600   
7807928    2018-01-24  6227874            1          324            30600   
8187531    2018-01-25  6214488            1          324            30600   
8187570    2018-01-25  6227874            1          324            30600   
8562619    2018-01-26  6214488            1          324            30600   
8562658    2018-01-26  6227874            1          324            30600   
8919336    2018-01-27  6215322            1          324            30600   
9188039    2018-01-28  6239843            1          324            30600   
9438982    2018-01-29  6239512            1          324            30600   
9438995    2018-01-29  6232358            1          324            30600   
9820698    2018-01-30  6232358            1          324            30600   
9820727    2018-01-30  6239512            1          324            30600   
10200233   2018-01-31  6239512            1          324            30600   
10200

In [11]:
df_trip_Jan.loc[df_trip_Jan['TRIPID'] == TRIPID_unique[45]].sort_values(by=['PROGRNUMBER'])

DAYOFSERVICE   TRIPID  PROGRNUMBER  STOPPOINTID  PLANNEDTIME_ARR  \
1414   2018-01-01  5959280            1          324            30600   
1527   2018-01-01  5959280            2          327            30701   
1649   2018-01-01  5959280            3         7113            30818   
1779   2018-01-01  5959280            4          127            30938   
1883   2018-01-01  5959280            5          112            31042   
1950   2018-01-01  5959280            6          113            31114   
1964   2018-01-01  5959280            7          114            31131   
2036   2018-01-01  5959280            8          115            31196   
2100   2018-01-01  5959280            9           37            31244   
2170   2018-01-01  5959280           10           38            31302   
2211   2018-01-01  5959280           11           39            31337   
2287   2018-01-01  5959280           12           40            31398   
2311   2018-01-01  5959280           13          146            31422   
2368   2018-01-01  5959280           14          147            31466   
2440   2018-01-01  5959280           15          184            31521   
2486   2018-01-01  5959280           16          185            31554   
2554   2018-01-01  5959280           17          186            31609   
2646   2018-01-01  5959280           18          187            31682   
2710   2018-01-01  5959280           19          188            31737   
2787   2018-01-01  5959280           20          189            31799   
2842   2018-01-01  5959280           21          190            31843   
2933   2018-01-01  5959280           22          191            31918   
3027   2018-01-01  5959280           23          819            31993   
3128   2018-01-01  5959280           24         7457            32069   
3255   2018-01-01  5959280           25         4725            32176   
3435   2018-01-01  5959280           26          273            32324   
3773   2018-01-01  5959280           27          405            32534   
3977   2018-01-01  5959280           28          408            32642   
4027   2018-01-01  5959280           29          409            32668   
4111   2018-01-01  5959280           30          410            32714   
4161   2018-01-01  5959280           31          411            32740   
4299   2018-01-01  5959280           32          412            32811   
4339   2018-01-01  5959280           33          413            32835   
4430   2018-01-01  5959280           34          414            32889   
4494   2018-01-01  5959280           35          415            32924   
4613   2018-01-01  5959280           36          416            32989   
4683   2018-01-01  5959280           37          417            33021   
4760   2018-01-01  5959280           38          418            33057   
4820   2018-01-01  5959280           39          419            33092   
4884   2018-01-01  5959280           40          420            33129   
4950   2018-01-01  5959280           41          421            33165   
5039   2018-01-01  5959280           42          422            33208   
5083   2018-01-01  5959280           43          423            33229   
5145   2018-01-01  5959280           44          424            33260   
5203   2018-01-01  5959280           45          425            33295   
5254   2018-01-01  5959280           46          426            33315   
5333   2018-01-01  5959280           47          427            33358   
5393   2018-01-01  5959280           48          428            33386   
5454   2018-01-01  5959280           49          429            33415   
5593   2018-01-01  5959280           50         6334            33487   
5751   2018-01-01  5959280           51         7660            33566   
5940   2018-01-01  5959280           52         3114            33655   
6014   2018-01-01  5959280           53         3115            33686   
6126   2018-01-01  5959280           54         3116           

In [12]:
df_trip_Jan['PLANNEDTIME_ARR'].max()

91217